<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>vGWAS Tutorial • vGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="vGWAS Tutorial">
<meta property="og:description" content="vGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2013.05.03</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/vGWAS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kullrich/vGWAS/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>vGWAS Tutorial</h1>
                        <h4 class="author">Xia Shen</h4>
            
            <h4 class="date">2020-04-15</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kullrich/vGWAS/blob/master/vignettes/vGWAS.Rmd"><code>vignettes/vGWAS.Rmd</code></a></small>
      <div class="hidden name"><code>vGWAS.Rmd</code></div>

    </div>

    
    
<div id="vignette---vgwas-tutorial" class="section level2">
<h2 class="hasAnchor">
<a href="#vignette---vgwas-tutorial" class="anchor"></a>Vignette - vGWAS Tutorial</h2>
<div id="a-brief-tutorial-of-the-r-package-vgwas" class="section level3">
<h3 class="hasAnchor">
<a href="#a-brief-tutorial-of-the-r-package-vgwas" class="anchor"></a>A Brief Tutorial of the R Package vGWAS</h3>
</div>
</div>
<div id="setup-notes" class="section level2">
<h2 class="hasAnchor">
<a href="#setup-notes" class="anchor"></a>Setup Notes</h2>
<p>To use the vGWAS package, of course, an R environment is required. Visit:</p>
<p><a href="http://www.r-project.org" class="uri">http://www.r-project.org</a></p>
<p>and install R for the operating system.</p>
<p>Start R and in the R console, type the following command to install the package dependencies first:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r">#install.packages("devtools")
#install.packages("hglm")
#install.packages("dglm")
#library(devtools)
#install_github("kullrich/vGWAS", build_vignettes = TRUE, dependencies = FALSE)</pre></body></html></div>
<p>Now the package is installed in the R library.</p>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>Load the library in the R console, and open this vignette with the command:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">vGWAS</span>)
<span class="co">#vignette("vGWAS")</span></pre></body></html></div>
<p>Six main functions in the package are ready to use:</p>
<ol style="list-style-type: decimal">
<li>brown.forsythe.test</li>
<li>vGWAS</li>
<li>vGWAS.gc</li>
<li>vGWAS.variance</li>
<li>plot (S3 method for vGWAS object)</li>
<li>summary (S3 method for vGWAS object)</li>
</ol>
<p>Run the following commands to load the example data:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">pheno</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">geno</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">chr</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">map</span>)</pre></body></html></div>
<p>pheno is a numeric vector of the simulated phenotypic values. By running:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">pheno</span>, <span class="kw">breaks</span> <span class="kw">=</span> <span class="fl">30</span>, <span class="kw">density</span> <span class="kw">=</span> <span class="fl">15</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="st">"darkred"</span>)</pre></body></html></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>The command:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">chr</span>)
<span class="co">#&gt; chr</span>
<span class="co">#&gt;    1    2    3    4    5 </span>
<span class="co">#&gt; 5000 3000 4000 2000 6000</span></pre></body></html></div>
<p>shows exactly the number of markers on each of the five simulated chromosomes.</p>
<p>Now, the objects loaded in R are ready for a vGWA scan, which can be done using the single command:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">vgwa</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/vGWAS.html">vGWAS</a></span>(<span class="kw">phenotype</span> <span class="kw">=</span> <span class="no">pheno</span>, <span class="kw">geno.matrix</span> <span class="kw">=</span> <span class="no">geno</span>, <span class="kw">marker.map</span> <span class="kw">=</span> <span class="no">map</span>, <span class="kw">chr.index</span> <span class="kw">=</span> <span class="no">chr</span>, <span class="kw">pB</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co">#vgwa &lt;- vGWAS(phenotype = pheno, geno.matrix = geno, marker.map = map, chr.index = chr, pB = TRUE)</span></pre></body></html></div>
<p>A progress bar will indicate the progress of the scan if pB is set to TRUE.</p>
<p>When the scan is finished, all the output statistics will be returned as a list into the object vgwa, which belongs to the class ‘vGWAS’. Any object that has a structure belonging to class ‘vGWAS’ can be directly passed into S3 method function plot. For instance, simply run the following command, we can plot the results in vgwa:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">vgwa</span>)
<span class="co">#&gt; nominal significance threshold with Bonferroni correction for 20000 tests are calculated.</span></pre></body></html></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>which produces Figure 2. There is a clear peak above the Bonferroni corrected threshold (dashed orange line).</p>
<p>Regarding the marker that gave the highest score, the heritability explained by the mean and variance can be split and calculated via:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/vGWAS.variance.html">vGWAS.variance</a></span>(<span class="kw">phenotype</span> <span class="kw">=</span> <span class="no">pheno</span>, <span class="kw">marker.genotype</span> <span class="kw">=</span> <span class="no">geno</span>[,<span class="no">vgwa</span>$<span class="no">p.value</span> <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">vgwa</span>$<span class="no">p.value</span>)])
<span class="co">#&gt; variance explained by the mean part of model:</span>
<span class="co">#&gt; 4.2 %</span>
<span class="co">#&gt; variance explained by the variance part of model:</span>
<span class="co">#&gt; 23.06 %</span>
<span class="co">#&gt; variance explained in total:</span>
<span class="co">#&gt; 27.26 %</span>
<span class="co">#&gt; $vm</span>
<span class="co">#&gt;         C </span>
<span class="co">#&gt; 0.1026317 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $vv</span>
<span class="co">#&gt;         C </span>
<span class="co">#&gt; 0.5628937 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ve</span>
<span class="co">#&gt;       C </span>
<span class="co">#&gt; 1.77584 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $vp</span>
<span class="co">#&gt;        C </span>
<span class="co">#&gt; 2.441366</span></pre></body></html></div>
<p>The output can also be stored if assigning the function call to an object.</p>
<p>To correct for population confounding, the package applies the method HEVA (h-likelihood-based efficient variance association), described in the reference paper of the package.</p>
<p>One may input pre-calculated kinship matrix, e.g. IBD or IBS matrix, and call vGWAS with heva = TRUE, kinship = calculated.kinship.</p>
<p>Instead, the inbuilt function vGWAS.heva automatically constructs a simple genomic kinship from the genotype data. Simply, runing:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">geno.coding</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">geno</span>), <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">geno</span>))
<span class="co">#pb &lt;- txtProgressBar(style = 3)</span>
<span class="kw">for</span> (<span class="no">j</span> <span class="kw">in</span> <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">geno</span>)) {
       <span class="no">geno.coding</span>[,<span class="no">j</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">geno</span>[,<span class="no">j</span>] <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">geno</span>[,<span class="no">j</span>]))[<span class="fl">1</span>])*<span class="fl">2</span> - <span class="fl">1</span>
       <span class="co">#setTxtProgressBar(pb, j/ncol(geno))</span>
}
<span class="fu"><a href="https://rdrr.io/r/graphics/image.html">image</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/crossprod.html">tcrossprod</a></span>(<span class="no">geno.coding</span>))</pre></body></html></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>creates a coded genotype matrix (contains -1 and 1 in this example) shown in Figure 3.</p>
<p>vGWAS.heva provides correction for phenotype and also visualization of population stratification. When calling vGWAS.heva or vGWAS as:</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co">#vgwa2 &lt;- vGWAS(phenotype = pheno, geno.matrix = geno.coding, heva = TRUE, marker.map = map, chr.index = chr)</span>
<span class="no">vgwa2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/vGWAS.gc.html">vGWAS.gc</a></span>(<span class="no">vgwa</span>)</pre></body></html></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Figure 4 is generated simultaneously with the calculation to visualize the stratification in the population via the first two principle components of the kinship matrix.</p>
<p>Plotting the this object as follows will generate a new vGWAS plot in Figure 5. In this example, the difference between the vGWA results with and without HEVA correction is small, but they might differ a lot for some datasets (See the reference paper in the package).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(<span class="no">vgwa2</span>)
<span class="co">#&gt; nominal significance threshold with Bonferroni correction for 20000 tests are calculated.</span></pre></body></html></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div id="remarks" class="section level2">
<h2 class="hasAnchor">
<a href="#remarks" class="anchor"></a>Remarks</h2>
<p>The package source and further development information are on the R-Forge project page: <a href="https://r-forge.r-project.org/projects/vgwas/" class="uri">https://r-forge.r-project.org/projects/vgwas/</a></p>
<p>A reimplementation of vGWAS as a devtool package can be found here: <a href="https://github.com/kullrich/vGWAS" class="uri">https://github.com/kullrich/vGWAS</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kristian Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
