<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>vGWAS basic tutorial â€¢ vGWAS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="vGWAS basic tutorial">
<meta property="og:description" content="vGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2024.10.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/vGWAS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vGWASsparse.html">vGWAS plink tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kullrich/vGWAS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>vGWAS basic tutorial</h1>
                        <h4 data-toc-skip class="author">Xia Shen</h4>
                        <h4 data-toc-skip class="author">Kristian
Ullrich<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
            
            <h4 data-toc-skip class="date">2024-10-23</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kullrich/vGWAS/blob/HEAD/vignettes/vGWAS.Rmd" class="external-link"><code>vignettes/vGWAS.Rmd</code></a></small>
      <div class="hidden name"><code>vGWAS.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="vignette---vgwas-basic-tutorial">Vignette - vGWAS basic tutorial<a class="anchor" aria-label="anchor" href="#vignette---vgwas-basic-tutorial"></a>
</h2>
<div class="section level3">
<h3 id="a-brief-tutorial-of-the-r-package-vgwas">A Brief Tutorial of the R Package vGWAS<a class="anchor" aria-label="anchor" href="#a-brief-tutorial-of-the-r-package-vgwas"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="setup-notes">Setup Notes<a class="anchor" aria-label="anchor" href="#setup-notes"></a>
</h2>
<p>To use the vGWAS package, of course, an R environment is required.
Visit:</p>
<p><a href="http://www.r-project.org" class="external-link uri">http://www.r-project.org</a></p>
<p>and install R for the operating system.</p>
<p>Start R and in the R console, type the following command to install
the package dependencies first:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#install.packages("devtools")</span></span>
<span><span class="co">#install.packages("knitr")</span></span>
<span><span class="co">#install.packages("dglm")</span></span>
<span><span class="co">#install.packages("doParallel")</span></span>
<span><span class="co">#install.packages("foreach")</span></span>
<span><span class="co">#install.packages("genio")</span></span>
<span><span class="co">#install.packages("hglm")</span></span>
<span><span class="co">#install.packages("Matrix")</span></span>
<span><span class="co">#install.packages("onewaytests")</span></span>
<span><span class="co">#library(devtools)</span></span>
<span><span class="co">#devtools::install_github("kullrich/vGWAS", build_vignettes = TRUE, dependencies = FALSE)</span></span></code></pre></div>
<p>Now the package is installed in the R library.</p>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>Load the library in the R console, and open this vignette with the
command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kullrich/vGWAS" class="external-link">vGWAS</a></span><span class="op">)</span></span></code></pre></div>
<p>Nine main functions in the package are ready to use:</p>
<ol style="list-style-type: decimal">
<li>bfmedian.test</li>
<li>brown.forsythe.tes</li>
<li>getMAF</li>
<li>vGWAS</li>
<li>vGWASparallel</li>
<li>vGWAS.gc</li>
<li>vGWAS.variance</li>
<li>plot (S3 method for vGWAS object)</li>
<li>summary (S3 method for vGWAS object)</li>
</ol>
<p>Run the following commands to load the example data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span></span></code></pre></div>
<p>pheno is a numeric vector of the simulated phenotypic values. By
running:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">pheno</span>, breaks <span class="op">=</span> <span class="fl">30</span>, density <span class="op">=</span> <span class="fl">15</span>, col <span class="op">=</span> <span class="st">"darkred"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>The command:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span>
<span><span class="co">#&gt; chr</span></span>
<span><span class="co">#&gt;    1    2    3    4    5 </span></span>
<span><span class="co">#&gt; 5000 3000 4000 2000 6000</span></span></code></pre></div>
<p>shows exactly the number of markers on each of the five simulated
chromosomes.</p>
<p>To pre-filter the <code>geno.matrix</code> data for
minor-allele-frequency, use the <code>getMAF</code> function as
follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]    93 20000</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMAF.html">getMAF</a></span><span class="op">(</span>geno.matrix <span class="op">=</span> <span class="va">geno</span>, geno.snp <span class="op">=</span> <span class="st">"col"</span><span class="op">)</span></span>
<span><span class="va">geno.maf</span> <span class="op">&lt;-</span> <span class="va">geno</span><span class="op">[</span>, <span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">geno.maf</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]    93 19075</span></span>
<span><span class="va">chr.maf</span> <span class="op">&lt;-</span> <span class="va">chr</span><span class="op">[</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="va">map.maf</span> <span class="op">&lt;-</span> <span class="va">map</span><span class="op">[</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">]</span></span></code></pre></div>
<p>Now, the objects loaded in R are ready for a vGWA scan, which can be
done using the single command:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vgwa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vGWAS.html">vGWAS</a></span><span class="op">(</span></span>
<span>  phenotype <span class="op">=</span> <span class="va">pheno</span>,</span>
<span>  geno.matrix <span class="op">=</span> <span class="va">geno</span>,</span>
<span>  marker.map <span class="op">=</span> <span class="va">map</span>,</span>
<span>  chr.index <span class="op">=</span> <span class="va">chr</span>,</span>
<span>  pB <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>A progress bar will indicate the progress of the scan if pB is set to
TRUE.</p>
<p>The same can be run using multiple cores.</p>
<p><strong>Note:</strong> using the <code>vGWASparallel</code> function
one need to specify, if the SNPs in the <code>geno.matrix</code> are at
columns (<code>geno.snp = "col"</code>) or at rows
(<code>geno.snp = "row"</code>)</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#vgwa &lt;- vGWASparallel(</span></span>
<span><span class="co">#  phenotype = pheno,</span></span>
<span><span class="co">#  geno.matrix = geno,</span></span>
<span><span class="co">#  geno.snp = "col",</span></span>
<span><span class="co">#  marker.map = map,</span></span>
<span><span class="co">#  chr.index = chr,</span></span>
<span><span class="co">#  pB = FALSE,</span></span>
<span><span class="co">#  ncores = 2)</span></span></code></pre></div>
<p>When the scan is finished, all the output statistics will be returned
as a list into the object vgwa, which belongs to the class â€˜vGWASâ€™. Any
object that has a structure belonging to class â€˜vGWASâ€™ can be directly
passed into S3 method function plot. For instance, simply run the
following command, we can plot the results in vgwa:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">vgwa</span><span class="op">)</span></span>
<span><span class="co">#&gt; nominal significance threshold with Bonferroni correction for 20000 tests are calculated.</span></span></code></pre></div>
<p><img src="vGWAS_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>which produces Figure 2. There is a clear peak above the Bonferroni
corrected threshold (dashed orange line).</p>
<p>Regarding the marker that gave the highest score, the heritability
explained by the mean and variance can be split and calculated via:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get marker with lowest p.value</span></span>
<span><span class="va">marker.lowest</span> <span class="op">&lt;-</span> <span class="va">vgwa</span><span class="op">$</span><span class="va">p.value</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">vgwa</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vGWAS.variance.html">vGWAS.variance</a></span><span class="op">(</span></span>
<span>  phenotype <span class="op">=</span> <span class="va">pheno</span>,</span>
<span>  marker.genotype <span class="op">=</span> <span class="va">geno</span><span class="op">[</span>, <span class="va">marker.lowest</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; variance explained by the mean part of model:</span></span>
<span><span class="co">#&gt; 4.2 %</span></span>
<span><span class="co">#&gt; variance explained by the variance part of model:</span></span>
<span><span class="co">#&gt; 23.06 %</span></span>
<span><span class="co">#&gt; variance explained in total:</span></span>
<span><span class="co">#&gt; 27.26 %</span></span>
<span><span class="co">#&gt; $vm</span></span>
<span><span class="co">#&gt;         C </span></span>
<span><span class="co">#&gt; 0.1026317 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $vv</span></span>
<span><span class="co">#&gt;         C </span></span>
<span><span class="co">#&gt; 0.5628937 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $ve</span></span>
<span><span class="co">#&gt;       C </span></span>
<span><span class="co">#&gt; 1.77584 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $vp</span></span>
<span><span class="co">#&gt;        C </span></span>
<span><span class="co">#&gt; 2.441366</span></span></code></pre></div>
<p>The output can also be stored if assigning the function call to an
object.</p>
</div>
<div class="section level2">
<h2 id="remarks">Remarks<a class="anchor" aria-label="anchor" href="#remarks"></a>
</h2>
<p>The package source and further development information are on the
R-Forge project page: <a href="https://r-forge.r-project.org/projects/vgwas/" class="external-link uri">https://r-forge.r-project.org/projects/vgwas/</a></p>
<p>A reimplementation of <code>vGWAS</code> introducing the parallel
function <code>vGWASparallel</code> as a devtool package can be found
here: <a href="https://github.com/kullrich/vGWAS" class="external-link uri">https://github.com/kullrich/vGWAS</a></p>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Max Planck Institute For Evolutionary Biology, <a href="mailto:ullrich@evolbio.mpg.de" class="email">ullrich@evolbio.mpg.de</a><a href="#fnref1" class="footnote-back">â†©ï¸Ž</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kristian Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
