<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Variance Genome-wide Association (parallel) — vGWASparallel • vGWAS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Variance Genome-wide Association (parallel) — vGWASparallel"><meta property="og:description" content="Variance Genome-wide association for using
nonparametric variance test and other"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vGWAS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2024.10.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/vGWAS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/vGWASsparse.html">vGWAS plink tutorial</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kullrich/vGWAS/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Variance Genome-wide Association (parallel)</h1>
    <small class="dont-index">Source: <a href="https://github.com/kullrich/vGWAS/blob/HEAD/R/vGWASparallel.R" class="external-link"><code>R/vGWASparallel.R</code></a></small>
    <div class="hidden name"><code>vGWASparallel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Variance Genome-wide association for using
nonparametric variance test and other</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">vGWASparallel</span><span class="op">(</span><span class="va">phenotype</span>, <span class="va">geno.matrix</span>, geno.snp <span class="op">=</span> <span class="st">"row"</span>,</span>
<span>marker.map <span class="op">=</span> <span class="cn">NULL</span>, chr.index <span class="op">=</span> <span class="cn">NULL</span>, method <span class="op">=</span> <span class="st">"bfmedian"</span>,</span>
<span>p.adjust.method <span class="op">=</span> <span class="st">"none"</span>, include.het <span class="op">=</span> <span class="cn">FALSE</span>, pB <span class="op">=</span> <span class="cn">TRUE</span>, ncores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>phenotype</dt>
<dd><p>a <code>numeric</code> or <code>logical</code> vector
of the phenotyic values.</p></dd>


<dt>geno.matrix</dt>
<dd><p>a <code>matrix</code> or <code>data.frame</code> or
<code>sparseMatrix</code> with individuals as columns and markers as rows
(geno.snp = "row") or individuals as rows and markers as columns
(geno.snp = "col").</p></dd>


<dt>marker.map</dt>
<dd><p>a <code>numeric</code> vector giving the marker map
positions for each chromosome.</p></dd>


<dt>chr.index</dt>
<dd><p>a <code>numeric</code> vector giving the chromosome index
for each marker.</p></dd>


<dt>geno.snp</dt>
<dd><p>if individuals at columns and markers at rows use "row" else
if individuals at rows and markers at columns use "col"</p></dd>


<dt>method</dt>
<dd><p>the test method to use (default = bfmedian).
Default is set to the Brown-Forsythe's Test of Equality of Variances
using group medians.
There are 31 other tests available via the onewaytests package:
Alvandi's F test ("af"),
Alexander-Govern test ("ag"),
Alvandi's generalized p-value ("agp"),
One-way analysis of variance ("aov"),
Approximate F test ("ap"),
Adjusted Welch's heteroscedastic F test ("aw"),
B square test ("b2"),
Brown-Forsythe test ("bf"),
Box F test ("box"),
Cochran test ("cochran"),
Generalized tests equivalent to Parametric Bootstrap ("gtb"),
Generalized tests equivalent to Fiducial tests ("gtf"),
Variance homogeneity tests ("homog"),
James second order test ("james"),
Johansen F test ("johansen"),
Kruskal-Wallis test ("kw"),
Modified Brown-Forsythe test ("mbf"),
Mann-Whitney U test ("mw"),
Anderson-Darling normility test ("nor_ad"),
Cramer-vin Mises normility test ("nor_cvm"),
Kolmogorov-Smirnov normility test ("nor_ks"),
Pearson Chi-square normility test ("nor_pct"),
Shapiro-Wilk normility test ("nor_sw"),
Shapiro-Francia normility test ("nor_sf"),
Permutation F test ("pf"),
Scott-Smith test ("ss"),
Student's t-test ("st"),
Welch-Aspin test ("wa"),
Welch's heteroscedastic F test with trimmed means and Winsorized variances
("welch"),
Weerahandi's generalized F test ("wgf"),
Welch's t-test ("wt").</p></dd>


<dt>test.na.rm</dt>
<dd><p>a logical value indicating whether NA values should be
stripped before the computation proceeds. Default us set to TRUE.</p></dd>


<dt>p.adjust.method</dt>
<dd><p>correction method (default = "none").
There are 8 p-value correction methods available via the p.adjust function:
"holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none"</p></dd>


<dt>include.het</dt>
<dd><p>specify if heterozygous calls should be split and added
equally to homozygous ref and alt counts (default = FALSE)</p></dd>


<dt>pB</dt>
<dd><p>show progress bar</p></dd>


<dt>ncores</dt>
<dd><p>number of cores to parallelize (default = 1)</p></dd>


<dt>test.alpa</dt>
<dd><p>the level of significance to assess the statistical
difference. Default is set to alpha = 0.05.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>data.frame</code> containing columns of <code>marker</code> names,
<code>chromosome</code> indices, <code>marker.map</code> positions,
test <code>statistic</code> values, and <code>p.value</code> for each position.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Shen, X., Pettersson, M., Ronnegard, L. and Carlborg, O.
(2011): <b>Inheritance beyond plain heritability:
variance-controlling genes in <em>Arabidopsis thaliana</em></b>.
<em>PLoS Genetics</em>, <b>8</b>, e1002839.<br></p>
<p>Ronnegard, L., Shen, X. and Alam, M. (2010):
<b>hglm: A Package for Fitting Hierarchical Generalized
Linear Models</b>. <em>The R Journal</em>, <b>2</b>(2), 20-28.<br></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="package-vGWAS.html">package-vGWAS</a></code> <code>onewaytests</code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Xia Shen</p>
<p>Kristian Ullrich</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># ----- load data ----- #</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">geno</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">chr</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># ----- variance GWA scan ----- #</span></span></span>
<span class="r-in"><span><span class="va">vgwa</span> <span class="op">&lt;-</span> <span class="fu">vGWASparallel</span><span class="op">(</span>phenotype <span class="op">=</span> <span class="va">pheno</span>, geno.matrix <span class="op">=</span> <span class="va">geno</span>,</span></span>
<span class="r-in"><span>marker.map <span class="op">=</span> <span class="va">map</span>, chr.index <span class="op">=</span> <span class="va">chr</span>,</span></span>
<span class="r-in"><span>geno.snp <span class="op">=</span> <span class="st">"col"</span>, pb <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in vGWASparallel(phenotype = pheno, geno.matrix = geno, marker.map = map,     chr.index = chr, geno.snp = "col", pb = FALSE):</span> unused argument (pb = FALSE)</span>
<span class="r-in"><span><span class="co"># ----- other test GWA scan ----- #</span></span></span>
<span class="r-in"><span><span class="va">vgwa.mw</span> <span class="op">&lt;-</span> <span class="fu">vGWASparallel</span><span class="op">(</span>phenotype <span class="op">=</span> <span class="va">pheno</span>, geno.matrix <span class="op">=</span> <span class="va">geno</span>,</span></span>
<span class="r-in"><span>marker.map <span class="op">=</span> <span class="va">map</span>, chr.index <span class="op">=</span> <span class="va">chr</span>,</span></span>
<span class="r-in"><span>geno.snp <span class="op">=</span> <span class="st">"col"</span>, method <span class="op">=</span> <span class="st">"mw"</span>, pb <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in vGWASparallel(phenotype = pheno, geno.matrix = geno, marker.map = map,     chr.index = chr, geno.snp = "col", method = "mw", pb = FALSE):</span> unused argument (pb = FALSE)</span>
<span class="r-in"><span><span class="co"># ----- multiple cores ----- #</span></span></span>
<span class="r-in"><span><span class="va">vgwa.st</span> <span class="op">&lt;-</span> <span class="fu">vGWASparallel</span><span class="op">(</span>phenotype <span class="op">=</span> <span class="va">pheno</span>, geno.matrix <span class="op">=</span> <span class="va">geno</span>,</span></span>
<span class="r-in"><span>marker.map <span class="op">=</span> <span class="va">map</span>, chr.index <span class="op">=</span> <span class="va">chr</span>,</span></span>
<span class="r-in"><span>geno.snp <span class="op">=</span> <span class="st">"col"</span>, method <span class="op">=</span> <span class="st">"st"</span>, ncores <span class="op">=</span> <span class="fl">2</span>, pb <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in vGWASparallel(phenotype = pheno, geno.matrix = geno, marker.map = map,     chr.index = chr, geno.snp = "col", method = "st", ncores = 2,     pb = FALSE):</span> unused argument (pb = FALSE)</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kristian Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

